### PR TITLE
Update TextEditor to a new workflow

### PR DESCRIPTION
**Pull request by irina060981**
For the issue #327 

Now text editor screen has the following steps:

image

**After clicking type/paste in** 
image

After inserting text (more than 5 characters) and loosing focus (click any other point on the screen out of the textarea) 
it will detect sourceType (text or tei), language and text direction (if sourceType === text)

image

**After clicking upload**
image

image

**After clicking New Alignment**
image

Updated labels from @abrasax feedback

image

### CODE REVIEW

File: src/lib/controllers/detect-text-controller.js
Comment (irina060981) [inline src/lib/controllers/detect-text-controller.js:6]: This is a controller for language detection.
It is used after text is updated.

File: src/lib/controllers/texts-controller.js
Comment (irina060981) [inline src/lib/controllers/texts-controller.js:84]: Used in a vue component to update visibility of poperties parts

File: src/lib/data/source-text.js
Comment (irina060981) [inline src/lib/data/source-text.js:38]: this is a flag for checking if language was already detected, to avoid double checks (to safe API requests)

File: src/lib/data/source-text.js
Comment (irina060981) [inline src/lib/data/source-text.js:84]: If a user clears text completly - the flag is set to false and detection would be done.

File: src/lib/data/source-text.js
Comment (irina060981) [inline src/lib/data/source-text.js:97]: These are minimal conditions for language detection

File: src/vue/text-editor/actions-menu-text-editor.vue
Comment (irina060981) [inline src/vue/text-editor/actions-menu-text-editor.vue:9]: I commented for now - because there is not fully clear would this button be used.

File: src/lib/data/source-text.js
Comment (zfletch) [reply src/lib/data/source-text.js:97]: Am I understanding the logic of this code correctly?

1. Once the text is longer than 5 characters, the language will be detected
2. The language will not be detected again unless the text length is 0

This seems like it could lead to problems for users who are typing text in the box. In this case, wouldn't we want to check the language when there is more data available? Or is the expected workflow that users change the language manually when typing text?

**Review by zfletch — CHANGES_REQUESTED** [review]
The logic required in https://github.com/alpheios-project/alignment-editor-new/pull/330 seems basic, but it surprisingly difficult. This is a great PR! I left a few suggestions. I am marking it as "Request changes" because of two of them I think are particularly important:

1. Checking in `package-lock.json`
2. Double-checking the regex and XML detection functionality to make sure it works as expected

File: src/lib/data/source-text.js
Comment (irina060981) [reply src/lib/data/source-text.js:97]: The logic here is the following:

- the event that will update current text length - is `blur`. It is done to reduce amount of text updates due to the fact, that text could be rather big.
- so the text would be updated and length would be calculated inside this text only after one of the follwoing events:
   - user unfocus the textarea
   - user uploaded the text from external resource
   
After a user removed the focus - the check would be done, and a user will be given an option to change language/direction and sourceType manually.

I set a limit to the 5 characters because if there are less characters - it is too few for lnguage detection.

@zfletch , in this case do you see additional risks?

File: package-lock.json
Comment (irina060981) [reply package-lock.json:1]: I deleted it manually.
Because I was updating the version of `alpheios-core`.
@zfletch what exactly risks do you see here?

File: src/lib/controllers/detect-text-controller.js
Comment (irina060981) [reply src/lib/controllers/detect-text-controller.js:31]: oh - I forgot to remove it :) thank you

File: src/lib/controllers/detect-text-controller.js
Comment (irina060981) [reply src/lib/controllers/detect-text-controller.js:16]: No, the sense of this method is another.
We support both - plain text and TEI XML.
On this stage - when a user enters a text - we don't check the ability of this text to be tokenized. (we would get it as a result from the remote tokenizer service).

This check is only for the language detection.
We could detect a language only for plain text (if I send an xml - it would be treated as a plain text).
That's why I use this check - to avoid TEI xml for detection.

After reading your comment I think - I would simplify the check and would simply check for `</tei>` tag using `indexOf` method.

File: src/lib/controllers/detect-text-controller.js
Comment (irina060981) [reply src/lib/controllers/detect-text-controller.js]: The regexp is simplified

File: src/lib/data/langs/langs.js
Comment (irina060981) [reply src/lib/data/langs/langs.js:5]: I don't agree here - because `rtlLangs` is a small array with maximum of 12 elements (amount of rtl langs).
It has a simple list of languages and nothing else.
There is no information for the object or Map here.

it looks like the following : [ 'ara', 'per', 'syr', .... ] - and that's all.
I understand your logic and could find a usage for such objects (and use them for detection dictionary) - but here it is redundant from my point of view.
I don't think that we should demonize `indexOf` and `includes` methods of the array and don't ever use it .

File: src/lib/data/langs/langs.js
Comment (irina060981) [reply src/lib/data/langs/langs.js:38]: my explanations in the next comment

File: src/lib/data/langs/langs.js
Comment (irina060981) [reply src/lib/data/langs/langs.js:4]: I agree that this code could be rather confusing, but let me give you some explanations:

this class has the following actions:

- it uploads once a big json file with available languages
- while it uploads this json file - it collects `allLangs` array for options configuration and `rtlLangs` array for direction definition; we collect them once and use several times in the code

`collectLangsData` - uploads languges and creates `allLangs` and `rtlLangs`
`all` - returns `allLangs` 
`defineDirection` - returns check based on `rtlLangs`

That's why - I check if these arrays are ready before a method will try to return the value

As for scoped variables - I like to use such an approach with static classes - that don't have an instance.
This way - one instance of the variable would be created and used. And such a static lass could be easily requested in any place in the code. I am not sure how it is correct in english - but I think of it as a `single instance class`

**Comment by irina060981** [comment]
@zfletch thank you for such an attentive review! Hope I gave all the necessary explanations. Would be glad to get some answers from you

**Comment by irina060981** [comment]
- updated labels upon @abrasax review
- fixed a bug with json upload from @monzug feedback
- removed additional step for upload choice upon @monzug and @abrasax review

**Comment by irina060981** [comment]
- reset local text options on each new alignment

File: src/lib/data/source-text.js
Comment (zfletch) [reply src/lib/data/source-text.js:95]: Thanks for the explanation! If the language detection code only runs on `blur`, I don't think there is much danger.

File: src/lib/data/langs/langs.js
Comment (zfletch) [reply src/lib/data/langs/langs.js:5]: Those explanations make sense, and I won't let this hold up approving the PR. But I still disagree about both points and I'd like to explain why below:

> `rtlLangs` is a small array with maximum of 12 elements

I think it is misleading to say that `rtLangs` is an array of 12 elements when talking about the code in this file. The array is not defined in this file, but calculated from `@/lib/data/langs/langs-list.json`. If the performance of the code in this file depends on the contents of that other file, the code is [tightly coupled](https://en.wikipedia.org/wiki/Coupling_(computer_programming)). This means that if `langs-list.json` changes, the performance of the code could change drastically. It also means that the developer or maintainer of this code has to be aware of the details of `langs-list.json`, increasing cognitive load.

> I don't think that we should demonize `indexOf` and `includes` methods of the array and don't ever use it.

Code largely serves two purposes: 1. it is a way of communicating with the machine and 2. it is a way to communicate your intention to other programmers. Even if there are no performance issues, the type of data structure communicates to other developers what the data structure is intended to be used for. In this case, I suggest using a `Map` because `rtlLangs` is used only to detect membership. Any programmer looking at the code will understand its intended use immediately if a `Map` is used. But a programmer will have to understand the internal workings of the rest of the code in this file to understand its use if an `array` is used.

File: src/lib/data/langs/langs.js
Comment (zfletch) [reply src/lib/data/langs/langs.js:4]: Thanks for explaining this to me, it makes more sense now. My question is: Don't you get the same functionality with [static properties](https://javascript.info/static-properties-methods)? I thought that's how static properties behaved, but I'm not an expert in JavaScript classes, so I could be mistaken.

---

> I check if these arrays are ready before a method will try to return the value

One design pattern that may be useful is putting the check in the collect method instead of in the part of the code that calls the method.

Instead of:

```javascript
static get all () {
  if (allLangs.length === 0) {
    allLangs = this.collectLangsData()
  }
  return allLangs
}

static defineDirection (langCode) {
  if (rtlLangs.length === 0) {
    allLangs = this.collectLangsData()
  }
  return rtlLangs.includes(langCode) ? 'rtl' : 'ltr'
}

static collectLangsData () {
  ...
}
```

You could write it like this:

```javascript
static get all () {
  return this.collectLangsData()
}

static defineDirection (langCode) {
  return collectRtlLangs().includes(langCode) ? 'rtl' : 'ltr'
}

static collectLangsData () {
  if (allLangs.length !== 0) { return allLangs }
  ...
}

static collectRtLangsData () {
  if (rtLangs.length !== 0) { return rtLangs }
  ...
}
```

The benefit of doing it this way is that the you limit the methods that need to be aware of the state, making the code easier to reason about, and making it simpler to add additional functionality.

File: package-lock.json
Comment (zfletch) [reply package-lock.json]: If you're interested in a detailed explanation, [this article](https://classic.yarnpkg.com/blog/2016/11/24/lockfiles-for-all/) is a good read.

The short answer is that removing the lockfile can cause a huge amount of pain if one of your dependencies changes. You'd potentially have to manually check every combination of every dependency in every matching version.

Comment (irina060981) [reply package-lock.json]: thank you - will read - package.lock is returned

File: src/lib/data/langs/langs.js
Comment (irina060981) [reply src/lib/data/langs/langs.js:38]: oh - I see your concerns - ok - will fix

File: src/lib/data/langs/langs.js
Comment (irina060981) [reply src/lib/data/langs/langs.js:4]: Thank you for the advice - I made similiar refactoring

**Pull request merged**
